File: app/schemas/transaction.py
from pydantic import BaseModel, field_validator
from datetime import datetime
from uuid import UUID
from app.models.transaction import MovementType, EntityContext


class TransactionCreate(BaseModel):
    user_id: UUID
    description: str | None = None
    amount: int  # in cents
    movement: MovementType
    category_id: int | None = None
    entity_type: EntityContext = EntityContext.PF
    source: str = "whatsapp"
    occurred_at: datetime | None = None


class TransactionRead(TransactionCreate):
    id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TransactionSummary(BaseModel):
    total_cash_in: int
    total_cash_out: int
    balance: int
    by_category: dict[str, int]

File: src/main/kotlin/ai/koalla/core/domain/Transaction.kt
package ai.koalla.core.domain

import ai.koalla.core.entity.EntityContext
import ai.koalla.core.entity.MovementType
import java.time.OffsetDateTime
import java.util.UUID

/**
 * Pure domain model for a financial transaction.
 * No JPA annotations — safe to use across all layers.
 */
data class Transaction(
    val id: UUID,
    val userId: UUID,
    val description: String?,
    val amount: Long,
    val movement: MovementType,
    val categoryId: Int?,
    val entityType: EntityContext,
    val source: String,
    val externalId: String?,
    val occurredAt: OffsetDateTime?,
    val createdAt: OffsetDateTime,
    val updatedAt: OffsetDateTime
)
